### ISSUE TITLE
kvnemesis: TestKVNemesisSingleNode* tests time out

### ISSUE THREAD

**Issue by pavelkalinnikov** [cockroachdb/cockroach#111352]
(found in https://github.com/cockroachdb/cockroach/issues/110551#issuecomment-1737181159)

`TestKVNemesisSingleNode_ReproposalChaos` times out. To reproduce on `master` around c704759:

```bash
$ dev test --verbose pkg/kv/kvnemesis --filter=TestKVNemesisSingleNode_ReproposalChaos -- --test_env=COCKROACH_RANDOM_SEED=-1416151305199981714
```

Bisect points at #110651.

Two steps are blocked on [Apply](https://github.com/cockroachdb/cockroach/blob/d6a9b485ab3ba7394adb85c6e2b78b6ab3304653/pkg/kv/kvnemesis/kvnemesis.go#L113), one of which is a [ClosureTxnOperation](https://github.com/cockroachdb/cockroach/blob/b3641ccd02d9ab0aa53cd5e53b709f27c13433aa/pkg/kv/kvnemesis/applier.go#L153), and another one is a [DelRange](https://github.com/cockroachdb/cockroach/blob/b3641ccd02d9ab0aa53cd5e53b709f27c13433aa/pkg/kv/kvnemesis/applier.go#L128). Both are blocked on [lockTableWaiterImpl.WaitOn](https://github.com/cockroachdb/cockroach/blob/8205b437a19d2efea025de9ef39bae1cbb299acd/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L156) for 4 minutes, so I assume there is some deadlock caused by the way they conflict.

Here is the content of the blocked steps (run in this order):

```
// blocked for 4 minutes
start: x.DelRange(ctx, tk(340099245535959704), tk(16963214484241583826), true /* @s10 */)

// in the meantime, this command ends
// note that the key is in the DelRange above
end: x.Put(ctx, tk(6360205682711281622), sv(9)) // on 1 Put at /Table/100/"5843fd5ee06d3fd6": injected error

// blocked for 4 minutes
// note that the Del key is in the DelRange too
// as well as the key in GetForUpdateSkipLocked
start: x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
  txn.SetIsoLevel(isolation.ReadCommitted)
  b := &kv.Batch{}
  b.GetForUpdateSkipLocked(tk(15321639564963858289))
  b.Del(tk(13939358664956423074) /* @s11 */)
  txn.CommitInBatch(ctx, b)
  return nil
})
```

@arulajmani says it can also be caused by `GetForUpdateSkipLocked` added in that PR.

Jira issue: CRDB-31868

**Comment by arulajmani** [comment]
> @arulajmani says it can also be caused by GetForUpdateSkipLocked added in that PR.

I'm not entirely sure why, but I think that's suspicious given the bisect pointed to my PR and the reproduction steps don't have any shared locking requests.

**Comment by pavelkalinnikov** [comment]
`TestKVNemesisSingleNode` fimes out the same way. Repro:

```bash
dev test --verbose pkg/kv/kvnemesis --filter=TestKVNemesisSingleNode -- --test_env=COCKROACH_RANDOM_SEED=89214452366778625
```

**Comment by miraradeva** [comment]
Wrt the git bisect, I think this test started timing out much more easily after shared locks were added, but if you go back before that, and bump up the ops per txn to 10, it still times out reliably within 1,000 runs. I haven't been able to git bisect it to a specific commit.

**Comment by pavelkalinnikov** [comment]
Tried a few commits in the past. The latest "bad" one I tried was `98566223~500` (minus 500 commits). Currently stressing on `98566223~1000`, still haven't failed after 1.5k runs.

**Comment by pavelkalinnikov** [comment]
Leaving to you, @arulajmani and @miraradeva. Not sure how far back we can bisect this, maybe at this point it makes sense to actually look in detail at the operations/transactions history that leads to the stall.

**Comment by miraradeva** [comment]
This commit is definitely the first place where the test becomes significantly slower: https://github.com/cockroachdb/cockroach/pull/105272.

**Comment by arulajmani** [comment]
@pavelkalinnikov, @miraradeva I wonder if this could be the cause of these timeouts -- https://github.com/cockroachdb/cockroach/pull/111554. 

It could be, given this started popping up after we merged shared locks support to KVNemesis. Essentially, if there is a deadlock, we won't detect it without my patch, and we would hang indefinitely. Ditto if there is some sort of liveness issue with one of the shared locking requests. 

This is purely conjecture, and I haven't had a chance to dig into this issue. I won't be able to dig into this for the next couple of weeks, as Im travelling, but maybe one of you can take this investigation over?

**Comment by pavelkalinnikov** [comment]
Thanks @arulajmani for taking a look before your travel. I am stressing `kvnemesis` tests with your patch now, to see if it helps, and will report to the PR. Either way, I will defer to @miraradeva / KV team to review the PR and/or further investigate.

**Comment by pavelkalinnikov** [comment]
> This commit is definitely the first place where the test becomes significantly slower: #105272.

In this PR, `TestKVNemesisSingleNode_ReproposalChaos` was introduced, and other 2 tests were extended with:

```
invalidLeaseAppliedIndexProb: 0.1,
injectReproposalErrorProb:    0.1,
```

This injects some failures into the flow. This makes the test slower because it now has to retry commands with some probabilities.

I've seen the test stalls at commits before this PR too.

**Comment by miraradeva** [comment]
@pavelkalinnikov you're right, it times out even before https://github.com/cockroachdb/cockroach/pull/105272. I see timeouts after this [commit](https://github.com/cockroachdb/cockroach/pull/103267/commits/5a53d25d40eb7cdf09b9a4911fee37d71d710664) and not before it. But it gets increasingly hard to get it to time out the more back in time I go, so I'm not 100% sure this is the culprit.

**Comment by pavelkalinnikov** [comment]
@nvanbenschoten Could behaviours introduced in #103267 cause the transaction stalls that we're seeing in this test?

**Comment by miraradeva** [comment]
@pavelkalinnikov Nathan and I are discussing in https://cockroachlabs.slack.com/archives/C04AFSL6E3C/p1695746540784329.

**Comment by nvanbenschoten** [comment]
@arulajmani I believe that this is a duplicate of https://github.com/cockroachdb/cockroach/issues/112608, but I'm not entirely sure. Could you take a look and close if we're certain that it is?

**Comment by nvanbenschoten** [comment]
Closing, as we believe that this has been resolved by recent fixes to kvnemesis.
